### PR TITLE
custom datasource processors continued

### PR DESCRIPTION
**Pull request by stwiname**
Changes:

- Add tests for project manifest with custom ds and fix relevant issues
- Add tests for ds service and fix relevant issues
- Add test for fetch service with custom ds
- Add build command in repo root
- Fix sandbox globals not being reset if there are errors

Todo:

- [x] Increase code reuse with sandbox and ds processor

### CODE REVIEW

File: packages/node/src/indexer/ds-processor.service.ts [hunk 1]
```diff
@@ -84,18 +56,19 @@ export class DsProcessorService {
   getDsProcessor<D extends string, T extends SubqlNetworkFilter>(
```
Comment (stwiname) [inline packages/node/src/indexer/ds-processor.service.ts:56]: I'm not sure if we should be checking that the ds is part of the project data sources from the constructor

File: packages/contract-processors/src/jsonfy.ts [hunk 2]
```diff
@@ -13,9 +13,19 @@ import {
...
   validate(ds: JsonfyDatasource): void {
+    if (ds.kind !== this.kind) {
+      throw new Error('ds kind doesnt match processor');
+    }
+
+    for (const handler of ds.mapping.handlers) {
+      if (!(handler.kind in this.handlerProcessors)) {
+        throw new Error(`ds kind ${handler.kind} not one of ${Object.keys(this.handlerProcessors).join(', ')}`);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (ianhe8x) [reply packages/contract-processors/src/jsonfy.ts:27]: Agree.

File: packages/node/src/indexer/ds-processor.service.ts (hunk 1 above)
Comment (ianhe8x) [reply packages/node/src/indexer/ds-processor.service.ts:56]: Don't get the full context of your question. If the check is async, we can't do it in constructor though.

File: packages/common/src/project/models.ts [hunk 3]
```diff
@@ -157,9 +174,11 @@ export class CustomDataSourceBase<
 > implements SubqlCustomDatasource<K, T, M>
 {
   @IsString()
+  @IsDefined()
```
Comment (ianhe8x) [inline packages/common/src/project/models.ts:177]: I think `@IsString()` covers `@IsDefined()`

File: packages/node/src/indexer/ds-processor.service.ts (hunk 1 above)
Comment (stwiname) [reply packages/node/src/indexer/ds-processor.service.ts:73]: Its possible to do something like this:

```ts
const project = {
  ...rest,
  dataSources: [ds1, ds2],
}

const ds3 = ...;

new DsProcessorService(project).getDsProcessor(ds3);
```

ds3 has nothing to do with the project

**Review by ianhe8x — APPROVED** [review]
lgtm

**Pull request merged**